### PR TITLE
fix #2458 add HTTP proxy support to UrlConnectionHttpClient

### LINKED ISSUE #2458: UrlConnectionHttpClient should have builder methods for setting a Proxy
## Describe the Feature
UrlConnectionHttpClient should provide a builder method so that we can programmatically set a Proxy to be used on the `URL.openConnection()` call, so that it is `URL.openConnection(<proxy>)`.

## Is your Feature Request related to a problem?
It is difficult to use the smaller UrlConnectionHttpClient with a custom proxy setup.

## Proposed Solution
Add a builder option to provide a Proxy instance. Another option would be to expose your own methods for setting up the proxy including TLS mutual auth for TLS based proxies since Java doesn't appear to support that nicely.

## Additional Context

- [ ] I may be able to implement this feature request

## Your Environment

* AWS Java SDK version used:
* JDK version used:
* Operating System and version:

### PR DESCRIPTION
**Pull request by MikeDombo**
## Motivation and Context
Resolves #2458. The goal of this change is to add full proxy support to the UrlConnectionHttpClient so that our project can remove the Apache HTTP Client and instead rely on Java builtins. This change must support HTTP and HTTPS proxies with username/password authorization.

## Modifications
Adds HTTP proxy support to the UrlConnectionHttpClient by using largely the same `ProxyConfiguration` class as is in the Apache HTTP Client. If the user configures a proxy using the builder, and the request isn't going to a non-proxy host, then the request will go through the application proxy using Java's builtin `Proxy` class.

## Testing
Manually tested using a local squid proxy via HTTP without authentication.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by debora-ito** [comment]
@MikeDombo can you please add unit tests?

File: http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java [hunk 1]
```diff
@@ -294,6 +328,11 @@ public void abort() {
          * when constructing the SSL context.
          */
         Builder tlsTrustManagersProvider(TlsTrustManagersProvider tlsTrustManagersProvider);
+
+        /**
+         * Configuration that defines how to communicate via an HTTP proxy.
+         */
+        Builder proxyConfiguration(ProxyConfiguration proxyConfiguration);
```
Comment (joviegas) [inline http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java:335]: Can we add [Builder with consumer](https://github.com/aws/aws-sdk-java-v2/blob/ab747a55116c45788c16390d8c2d280bb063bc01/http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java#L304) also ?

File: http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java [hunk 2]
```diff
@@ -70,12 +75,16 @@
 
     private static final Logger log = Logger.loggerFor(UrlConnectionHttpClient.class);
     private static final String CLIENT_NAME = "UrlConnection";
+    private static final AttributeMap.Key<ProxyConfiguration> PROXY_CONFIGURATION =
```
Comment (joviegas) [inline http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java:78]: To be consistent with other http client lets assign the ProxyConfiguration directly to the class Field as done in other client constructors , I didnot get the benefit of assigning it to a AttributeMap and then doing a get.

Can we do something similar to [ApacheHttpClient](https://github.com/aws/aws-sdk-java-v2/blob/b673b8456a5e6c38132d5764661f64b1dcb981a6/http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java#L453) ?

File: http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java [hunk 3]
```diff
@@ -148,7 +157,17 @@ private HttpURLConnection createAndConfigureConnection(HttpExecuteRequest reques
     }
 
     private HttpURLConnection createDefaultConnection(URI uri, SSLSocketFactory socketFactory) {
-        HttpURLConnection connection = invokeSafely(() -> (HttpURLConnection) uri.toURL().openConnection());
+        HttpURLConnection connection;
```
Comment (joviegas) [inline http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java:160]: can we simplify the code as below ?

```java
    private HttpURLConnection createDefaultConnection(URI uri, SSLSocketFactory socketFactory) {

        Optional<Proxy> proxy = determineProxy(uri);

        HttpURLConnection connection = !proxy.isPresent() ?
                                       invokeSafely(() -> (HttpURLConnection) uri.toURL().openConnection())
                                                          :
                                       invokeSafely(() -> (HttpURLConnection) uri.toURL().openConnection(proxy.get()));

        if (connection instanceof HttpsURLConnection) {
            HttpsURLConnection httpsConnection = (HttpsURLConnection) connection;

            if (options.get(SdkHttpConfigurationOption.TRUST_ALL_CERTIFICATES)) {
                httpsConnection.setHostnameVerifier(NoOpHostNameVerifier.INSTANCE);
            }
            httpsConnection.setSSLSocketFactory(socketFactory);
        }

        if (proxy.isPresent() && shouldProxyAuthorize()){
            connection.addRequestProperty("proxy-authorization", String.format("Basic %s", encodedAuthToken(proxyConfiguration)));
        }
        connection.setConnectTimeout(saturatedCast(options.get(SdkHttpConfigurationOption.CONNECTION_TIMEOUT).toMillis()));
        connection.setReadTimeout(saturatedCast(options.get(SdkHttpConfigurationOption.READ_TIMEOUT).toMillis()));

        return connection;
    }

    private boolean shouldProxyAuthorize() {
        return this.proxyConfiguration != null 
               && StringUtils.isEmpty(this.proxyConfiguration.username()) 
               && StringUtils.isEmpty(this.proxyConfiguration.password());
    }


    /**
     * If a proxy is configured with username+password, then set the proxy-authorization header to authorize ourselves with the
     * proxy
     */
    private static String encodedAuthToken(ProxyConfiguration proxyConfiguration) {

        String authToken = String.format("%s:%s", proxyConfiguration.username(), proxyConfiguration.password());
        return Base64.getEncoder().encodeToString(authToken.getBytes(StandardCharsets.UTF_8));
    }

    private Optional<Proxy> determineProxy(URI uri) {
        if (isProxyEnabled(proxyConfiguration) && !isProxyHostExcluded(uri)) {
            return Optional.of(
                new Proxy(Proxy.Type.HTTP,
                          new InetSocketAddress(this.proxyConfiguration.host(), this.proxyConfiguration.port())));
        }
        return Optional.empty();
    }
```

**Comment by debora-ito** [comment]
Superseded by https://github.com/aws/aws-sdk-java-v2/pull/3112. Closing.

**Pull request closed without merging**